### PR TITLE
Update: add `indent` options for array and object literals (fixes #7473)

### LINKED ISSUE #7473: Version 3.9.0 broke multiline array indentation
**Tell us about your environment**
- **ESLint Version:** v3.9.0
- **Node Version:** v7.0.0-pre
- **npm Version:** 3.10.8

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

```
{
    "rules": {
        "indent": [
            "error",
            4
        ]
    }
}
```

**What did you do? Please include the actual source code causing the issue.**

``` js
var array = [1,
             2,
             3];
```

**What did you expect to happen?**

This was considered valid before this release. It would be great to have the `first` option for arrays too.

**What actually happened? Please include the actual, raw output from ESLint.**

```
2:14  error  Expected indentation of 4 spaces but found 13  indent
3:14  error  Expected indentation of 4 spaces but found 13  indent
```

### PR DESCRIPTION
**Pull request by not-an-aardvark**
**What is the purpose of this pull request? (put an "X" next to item)**

[x] Changes an existing rule

See https://github.com/eslint/eslint/issues/7473

**What changes did you make? (Give an overview)**

This adds an `ArrayExpression` and `ObjectExpression` option to the `indent` rule, as described in https://github.com/eslint/eslint/issues/7473.

**Is there anything you'd like reviewers to focus on?**

Nothing in particular.

### CODE REVIEW

File: lib/rules/indent.js [hunk 1]
```diff
@@ -694,27 +726,27 @@ module.exports = {
...
-                            parent.type === "CallExpression" ||
-                            parent.type === "ArrowFunctionExpression" ||
-                            parent.type === "NewExpression" ||
-                            parent.type === "LogicalExpression"
-                        ) {
-                            nodeIndent = nodeIndent + indentSize;
+                        } else if (parent.type === "ObjectExpression" || parent.type === "ArrayExpression") {
+                            nodeIndent += options[parent.type] * indentSize;
+                        } else if (parent.type === "CallExpression" || parent.type === "NewExpression") {
+                            nodeIndent += (typeof options.CallExpression.arguments === "number" ? options.CallExpression.arguments : 1) * indentSize;
```
Comment (mysticatea) [inline lib/rules/indent.js]: Incidentally, is `typeof` checks unnecessary for `ObjectExpression` differently to `CallExpression`?

File: lib/rules/indent.js [hunk 2]
```diff
@@ -694,27 +726,27 @@ module.exports = {
...
-                elementsIndent = nodeIndent + indentSize;
-
                 checkFirstNodeLineIndent(node, nodeIndent);
             } else {
                 nodeIndent = getNodeIndent(node).goodChar;
-                elementsIndent = nodeIndent + indentSize;
+            }
+
+            if (options[node.type] === "first") {
+                elementsIndent = elements[0].loc.start.column;
```
Comment (mysticatea) [inline lib/rules/indent.js]: This seems to throw `TypeError` if there are empty objects/arrays.

**Comment by larionov** [comment]
I tried to make it work with with indentation like this but had no success. Do you think this is even possible?
```
export default angular.module('foo', [
        module1,
        module2,
        module3
    ])
    .directive('bar', barDirective);
```
I have a rule set like this, and tried different values of ArrayExpression and MemberExpression, no luck.

`"indent": [2, 4, { "SwitchCase": 1, "MemberExpression": 1, "ArrayExpression": 1}]`

**Comment by not-an-aardvark** [comment]
@larionov The correct indentation for that case would be:

```js
export default angular.module('foo', [
        module1,
        module2,
        module3
])
    .directive('bar', barDirective);
```

or alternatively you could put the .directive call on the same line:

```js
export default angular.module('foo', [
        module1,
        module2,
        module3
]).directive('bar', barDirective);
```

Since the array after `foo` is opened on a line with an indentation of 0, it should also be closed on a line with an indentation of 0.

**Comment by larionov** [comment]
@not-an-aardvark  I use http://jsbeautifier.org/ plugin to format my code automatically and it formats this code like in my first example, so I thought that was somewhat generally accepted. But anyway I already formatted all the code like in your first example, not sure how will second one will go with huge lists of directives.

**Comment by mysticatea** [comment]
Just linked to #4081 since it's similar topic.

**Comment by NicholasBoll** [comment]
This PR seems to have broken `ObjectExpression` default. It used to be `1` and now it is `"first"`. So either the documentation is wrong (no longer default `1`) or the code is wrong. We had our CI suddenly fail with indentation errors with object expressions shortly after the `3.12.0` release.

Our configuration for indent is:
```js
"indent" : [2, 2, {"SwitchCase": 1}],
```

**Comment by gyandeeps** [comment]
@NicholasBoll Can you please open a new issue with what u r seeing. Thanks

**Pull request merged**